File: compatibility-tests/src/test/java/nnl/rocks/kactoos/list/JoinedTest.java
/**
 * The MIT License (MIT)
 * <p>
 * Copyright (c) 2017-2018 Yegor Bugayenko
 * <p>
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * <p>
 * The above copyright notice and this permission notice shall be included
 * in all copies or substantial portions of the Software.
 * <p>
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
package nnl.rocks.kactoos.list;

import org.hamcrest.MatcherAssert;
import org.hamcrest.Matchers;
import org.junit.Test;

/**
 * Test case for {@link nnl.rocks.kactoos.collection.Joined}.
 *
 * @author Vseslav Sekorin (vssekorin@gmail.com)
 * @version $Id: eef0d66faee6f1fd8dfedc6ae0badb3bb4e5dcd2 $
 * @since 0.20
 * @checkstyle JavadocMethodCheck (500 lines)
 * @checkstyle MagicNumber (500 line)
 */
@SuppressWarnings("PMD.TooManyMethods")
public final class JoinedTest {

    @Test
    public void behavesAsCollection() {
        MatcherAssert.assertThat(
            "Can't behave as a list",
            new Joined<Integer>(
                new ListOf<>(1, 2),
                new ListOf<>(3, 4)
            ),
            new BehavesAsList<>(2)
        );
    }

    @Test
    public void size() {
        MatcherAssert.assertThat(
            new Joined<Integer>(
                new ListOf<>(1, 2),
                new ListOf<>(3, 4)
            ).size(),
            Matchers.equalTo(4)
        );
    }

    @Test
    public void isEmpty() {
        MatcherAssert.assertThat(
            new Joined<Integer>(
                new ListOf<Integer>(5, 6)
            ).isEmpty(),
            Matchers.equalTo(false)
        );
    }

    @Test
    public void contains() {
        final int element = 0;
        MatcherAssert.assertThat(
            new Joined<Integer>(
                new ListOf<>(7, 8),
                new ListOf<>(9, element)
            ).contains(element),
            Matchers.equalTo(true)
        );
    }

    @Test
    public void iterator() {
        final String element = "element";
        MatcherAssert.assertThat(
            new Joined<String>(
                new ListOf<>(element, "first"),
                new ListOf<>("second", "third")
            ).iterator().next(),
            Matchers.equalTo(element)
        );
    }

    @Test
    public void toArray() {
        MatcherAssert.assertThat(
            new Joined<Integer>(
                new ListOf<>(11, 12),
                new ListOf<>(13, 14)
            ).toArray(),
            Matchers.equalTo(new ListOf<>(11, 12, 13, 14).toArray())
        );
    }

    @Test(expected = UnsupportedOperationException.class)
    public void add() {
        new Joined<String>(new ListOf<String>("add0")).add("new add0");
    }

    @Test(expected = UnsupportedOperationException.class)
    public void remove() {
        new Joined<String>(new ListOf<String>("remove")).remove("new remove");
    }

    @Test
    public void containsAll() {
        final String first = "item1";
        final String second = "item2";
        MatcherAssert.assertThat(
            new Joined<String>(
                new ListOf<>(first, "item3"),
                new ListOf<>(second, "item4")
            ).containsAll(new ListOf<>(first, second)),
            Matchers.equalTo(true)
        );
    }

    @Test(expected = UnsupportedOperationException.class)
    public void addAll() {
        new Joined<String>(
            new ListOf<String>("addAll")
        ).addAll(new ListOf<>("new addAll"));
    }

    @Test(expected = UnsupportedOperationException.class)
    public void addAllSecond() {
        new Joined<String>(
            new ListOf<String>("addAll1")
        ).addAll(22, new ListOf<>("new addAll1"));
    }

    @Test(expected = UnsupportedOperationException.class)
    public void removeAll() {
        new Joined<String>(
            new ListOf<String>("removeAll")
        ).removeAll(new ListOf<>("new removeAll"));
    }

    @Test(expected = UnsupportedOperationException.class)
    public void retainAll() {
        new Joined<String>().retainAll(new ListOf<>("retain", "All"));
    }

    @Test(expected = UnsupportedOperationException.class)
    public void clear() {
        new Joined<String>(new ListOf<String>("clear")).clear();
    }

    @Test
    public void get() {
        final String element = "element2";
        MatcherAssert.assertThat(
            new Joined<String>(
                new ListOf<>("element1"),
                new ListOf<>(element, "element3")
            ).get(1),
            Matchers.equalTo(element)
        );
    }

    @Test(expected = UnsupportedOperationException.class)
    public void set() {
        new Joined<String>(new ListOf<String>("set")).set(33, "new set");
    }

    @Test(expected = UnsupportedOperationException.class)
    public void addSecond() {
        new Joined<String>(new ListOf<String>("add")).add(44, "new add");
    }

    @Test(expected = UnsupportedOperationException.class)
    public void removeSecond() {
        new Joined<String>().remove(55);
    }

    @Test(expected = IndexOutOfBoundsException.class)
    public void listIteratorSecond() {
        new Joined<Integer>().listIterator(66);
    }

    @Test
    public void subList() {
        final String element = "elem2";
        MatcherAssert.assertThat(
            new Joined<String>(
                new ListOf<>("elem1", element),
                new ListOf<>("elem3", "elem4")
            ).subList(1, 3).iterator().next(),
            Matchers.equalTo(element)
        );
    }
}

File: kactoos-jvm/src/test/kotlin/nnl/rocks/kactoos/list/JoinedTest.kt

package nnl.rocks.kactoos.list

import org.hamcrest.MatcherAssert
import org.hamcrest.Matchers
import org.junit.Test

/**
 * Test case for [nnl.rocks.kactoos.collection.Joined].
 *
 * @author Vseslav Sekorin (vssekorin@gmail.com)
 * @version $Id: eef0d66faee6f1fd8dfedc6ae0badb3bb4e5dcd2 $
 * @since 0.20
 * @checkstyle JavadocMethodCheck (500 lines)
 * @checkstyle MagicNumber (500 line)
 */
class JoinedTest {

    @Test
    @Throws(Exception::class)
    fun behavesAsCollection() {
        MatcherAssert.assertThat(
                "Can't behave as a list",
                Joined(
                        ListOf(1, 2),
                        ListOf(3, 4)
                ),
                BehavesAsList(2)
        )
    }

    @Test
    @Throws(Exception::class)
    fun size() {
        MatcherAssert.assertThat(
                Joined(
                        ListOf(1, 2),
                        ListOf(3, 4)
                ).size,
                Matchers.equalTo(4)
        )
    }

    @Test
    @Throws(Exception::class)
    fun isEmpty() {
        MatcherAssert.assertThat(
                Joined(
                        ListOf(5, 6)
                ).isEmpty(),
                Matchers.equalTo(false)
        )
    }

    @Test
    @Throws(Exception::class)
    fun contains() {
        val element = 0
        MatcherAssert.assertThat(
                Joined(
                        ListOf(7, 8),
                        ListOf(9, element)
                ).contains(element),
                Matchers.equalTo(true)
        )
    }

    @Test
    @Throws(Exception::class)
    operator fun iterator() {
        val element = "element"
        MatcherAssert.assertThat(
                Joined(
                        ListOf(element, "first"),
                        ListOf("second", "third")
                ).iterator().next(),
                Matchers.equalTo(element)
        )
    }

    @Test
    @Throws(Exception::class)
    fun toArray() {
        MatcherAssert.assertThat<Array<Any>>(
                Joined(
                        ListOf(11, 12),
                        ListOf(13, 14)
                ).toTypedArray(),
                Matchers.equalTo<Array<Any>>(ListOf(11, 12, 13, 14).toTypedArray())
        )
    }

    @Test(expected = UnsupportedOperationException::class)
    @Throws(Exception::class)
    fun add() {
        Joined(ListOf("add0")).add("new add0")
    }

    @Test(expected = UnsupportedOperationException::class)
    @Throws(Exception::class)
    fun remove() {
        Joined(ListOf("remove")).remove("new remove")
    }

    @Test
    @Throws(Exception::class)
    fun containsAll() {
        val first = "item1"
        val second = "item2"
        MatcherAssert.assertThat(
                Joined(
                        ListOf(first, "item3"),
                        ListOf(second, "item4")
                ).containsAll(ListOf(first, second)),
                Matchers.equalTo(true)
        )
    }

    @Test(expected = UnsupportedOperationException::class)
    @Throws(Exception::class)
    fun addAll() {
        Joined(
                ListOf("addAll")
        ).addAll(ListOf("new addAll"))
    }

    @Test(expected = UnsupportedOperationException::class)
    @Throws(Exception::class)
    fun addAllSecond() {
        Joined(
                ListOf("addAll1")
        ).addAll(22, ListOf("new addAll1"))
    }

    @Test(expected = UnsupportedOperationException::class)
    @Throws(Exception::class)
    fun removeAll() {
        Joined(
                ListOf("removeAll")
        ).removeAll(ListOf("new removeAll"))
    }

    @Test(expected = UnsupportedOperationException::class)
    @Throws(Exception::class)
    fun retainAll() {
        Joined<String>().retainAll(ListOf("retain", "All"))
    }

    @Test(expected = UnsupportedOperationException::class)
    @Throws(Exception::class)
    fun clear() {
        Joined(ListOf("clear")).clear()
    }

    @Test
    @Throws(Exception::class)
    fun get() {
        val element = "element2"
        MatcherAssert.assertThat(
                Joined(
                        ListOf("element1"),
                        ListOf(element, "element3")
                )[1],
                Matchers.equalTo(element)
        )
    }

    @Test(expected = UnsupportedOperationException::class)
    @Throws(Exception::class)
    fun set() {
        Joined(ListOf("set")).set(33, "new set")
    }

    @Test(expected = UnsupportedOperationException::class)
    @Throws(Exception::class)
    fun addSecond() {
        Joined(ListOf("add")).add(44, "new add")
    }

    @Test(expected = UnsupportedOperationException::class)
    @Throws(Exception::class)
    fun removeSecond() {
        Joined<String>().removeAt(55)
    }

    @Test(expected = IndexOutOfBoundsException::class)
    @Throws(Exception::class)
    fun listIteratorSecond() {
        Joined<Int>().listIterator(66)
    }

    @Test
    @Throws(Exception::class)
    fun subList() {
        val element = "elem2"
        MatcherAssert.assertThat(
                Joined(
                        ListOf("elem1", element),
                        ListOf("elem3", "elem4")
                ).subList(1, 3).iterator().next(),
                Matchers.equalTo(element)
        )
    }
}
